### PR TITLE
feat: primitive rendering system

### PR DESCRIPTION
**Pull request by dalkia**
Completes #34

- Introduces a `BoxFactory`, `SphereFactory` and `PlaneFactory` to generate primitive meshes. The same cleanup used in the Colliders PR was applied on those Factories (using pools for temporary buffers generation)
- Introduces the `InstantiatePrimitiveRendererSystem` which adds primitive rendering to ECS world. Introduces `PrimitiveRendererComponent` and `PrimitiveMeshComponent` to achieve this. Unlike the Collider system, there is no mesh parameter change possible; since it will mean rebuilding the whole mesh.
- Introduces a 'ReleaseOutdatedRenderingSystem' to handle changes. The rebuilding of the whole mesh is handled in this system. If there is a change type, the whole mesh is rebuilt. 
- Create a pool for `Mesh` handling
- Modified the `UnityComponentPool` to handle the MeshRenderer cases. This are special, since MeshRenderer always goes with MeshFilter, so special attention should be handled there. 

Out of scope

- Material handling. For this PR, a new material is instantiated everytime the rendering is being used

### CODE REVIEW

File: Explorer/Assets/Scripts/CrdtEcsBridge/Components/Defaults/PBMeshRendererDefaults.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+using DCL.ECSComponents;
+
+namespace CrdtEcsBridge.Components.Defaults
+{
+    public static class PBMeshRendererDefaults
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/CrdtEcsBridge/Components/Defaults/PBMeshRendererDefaults.cs:5]: Let's agree on where we place defaults. I did it in the "feature folder" for another feature as they refer to the specific area. I propose to move them, WHYT?

File: Explorer/Assets/Scripts/Utility/Primitives/PrimitivesSize.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+using UnityEngine;
+
+namespace Utility.Primitives
+{
+    public static class PrimitivesSize
+    {
+        public static readonly int CUBE_SIZE = 1;
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/Primitives/PrimitivesSize.cs:7]: It should be `const`, not `static readonly`

File: Explorer/Assets/Scripts/Utility/Primitives/PrimitivesSize.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+        public static readonly Vector3 PLANE_SIZE = new (1, 1, 0.01f);
+        public static readonly float SPHERE_RADIUS = 0.5f;
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/Primitives/PrimitivesSize.cs:9]: should be `const`

File: Explorer/Assets/Scripts/Utility/Primitives/PlaneFactory.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
+using UnityEngine;
+
+namespace Utility.Primitives
+{
+    public static class PlaneFactory
+    {
+        public const int VERTICES_NUM = 8;
+        public const int TRIS_NUM = 12;
+        public static readonly Vector3 SIZE = PrimitivesSize.PLANE_SIZE;
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/Primitives/PlaneFactory.cs:9]: It does not make sense to keep an additional constant, just use `PrimitivesSize` directly

File: Explorer/Assets/Scripts/Utility/Primitives/PrimitivesSize.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+using UnityEngine;
+
+namespace Utility.Primitives
+{
+    public static class PrimitivesSize
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/Primitives/PrimitivesSize.cs:5]: Please move cylinder height here, let's keep everything in one place

File: Explorer/Assets/Scripts/Utility/Primitives/PrimitivesUtility.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Utility.Primitives
+{
+    public class PrimitivesUtility
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/Primitives/PrimitivesUtility.cs:6]: Make this class static

File: Explorer/Assets/Scripts/Utility/Primitives/PrimitivesUtility.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public static Vector2[] FloatArrayToV2List(IList<float> uvs)
+        {
+            var uvsResult = new Vector2[uvs.Count / 2];
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/Primitives/PrimitivesUtility.cs:10]: We already pool `UVs` in the corresponding primitives' factories. It is incorrect to replace them with a newly allocated instance

Comment (mikhail-dcl) [reply Explorer/Assets/Scripts/Utility/Primitives/PrimitivesUtility.cs:10]: You can consider every `new()` for arrays banned

File: Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+
+        /// <summary>
+        ///     It is either called when there is no collider or collider was invalidated before (set to null)
+        /// </summary>
+        private void Instantiate(ISetupMesh meshSetup, ref PrimitiveRendererComponent rendererComponent, ref PrimitiveMeshComponent meshComponent,
+            PBMeshRenderer sdkComponent, ref TransformComponent transformComponent)
+        {
+            MeshRenderer meshRendererGo = rendererPoolRegistry.Get();
+            meshRendererGo.material = new Material(Shader.Find("Universal Render Pipeline/Lit"));
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs:86]: You should create the material only once on the system.Initialize() and reuse it as shared material! This is why you get `SRPBatcher` pressure

File: Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+            Mesh mesh = meshPoolRegistry.Get();
+
+            meshSetup.Execute(sdkComponent, mesh);
+
+            meshComponent.Mesh = mesh;
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs:92]: What's the idea behind the second level of the pool in the form of `Mesh` pool? For which cases do you think it won't be 1 to 1 with the `MeshRenderer` pool?

File: Explorer/Assets/Scripts/Utility/Primitives/PlaneFactory.cs [hunk 10]
```diff
@@ -0,0 +1,83 @@
+using UnityEngine;
+
+namespace Utility.Primitives
+{
+    public static class PlaneFactory
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/Primitives/PlaneFactory.cs:5]: Consider introducing tests for each primitive factory similarly to `CylinderVariantsShould`: it's important and straightforward to validate the pure math logic and buffers reusing

File: Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            meshComponent.SDKType = sdkComponent.MeshCase;
+
+            rendererComponent.MeshRenderer = meshRendererGo;
+            rendererComponent.MeshFilter = meshRendererGo.GetComponent<MeshFilter>();
+            rendererComponent.MeshFilter.sharedMesh = mesh;
+
+            Transform rendererTransform = meshRendererGo.transform;
+            rendererTransform.SetParent(transformComponent.Transform, false);
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs:100]: Let's introduce extensions for component parenting: it's the same for `Collider` and `Renderer` and will be repeated in the future systems

File: Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+            InstantiateNonExistingRendererQuery(World);
+            TrySetupExistingRendererQuery(World);
+        }
+
+        [Query]
+        [All(typeof(PBMeshRenderer), typeof(SDKTransform), typeof(TransformComponent))]
+        [None(typeof(PrimitiveRendererComponent))]
+        private void InstantiateNonExistingRenderer(in Entity entity, ref PBMeshRenderer sdkComponent, ref TransformComponent transform)
+        {
+            var rendererComponent = new PrimitiveRendererComponent();
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs:52]: In general, it's fine but I would like to warn you that `PrimitiveRendererComponent` and `PrimitiveMeshComponent` always come hand in hand, it's a signal that they should be a part of the same component. You still can have different structures if you want (composition)

File: Explorer/Assets/Scripts/Utility/Primitives/SphereFactory.cs [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+            for (var lat = 0; lat < nbLat; lat++)
+            {
+                for (var lon = 0; lon <= nbLong; lon++)
+                {
+                    uvs[lon + (lat * (nbLong + 1)) + 1] =
+                        new Vector2(1f - ((float)lon / nbLong), (float)(lat + 1) / (nbLat + 1));
+                }
+            }
+
+            //uvs[lon + lat * (nbLong + 1) + 1] = new Vector2( (float)lon / nbLong, 1f - (float)(lat+1) / (nbLat+1) );
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/Primitives/SphereFactory.cs:70]: Let's remove this comment

File: Explorer/Assets/Scripts/Utility/Primitives/SphereFactory.cs [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+                    triangles[i++] = current;
+                    triangles[i++] = current + 1;
+                    triangles[i++] = next + 1;
+
+                    triangles[i++] = current;
+                    triangles[i++] = next + 1;
+                    triangles[i++] = next;
+
+                    if (current >= 402) { Debug.Log("AQUI HAY ERROR"); }
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/Utility/Primitives/SphereFactory.cs:105]: Let's remove these weird logs

File: Explorer/Assets/Scripts/ECS/ComponentsPooling/UnityComponentPool.cs [hunk 15]
```diff
@@ -11,11 +13,14 @@ public class UnityComponentPool<T> : IComponentPool<T> where T: Component
         private readonly Transform parentContainer;
         private readonly Transform rootContainer;
 
-        public UnityComponentPool(Transform rootContainer, int maxSize = 2048)
+        private Action<T> OnRelease;
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/ComponentsPooling/UnityComponentPool.cs:16]: it should be `readonly` and starts with a small letter, right?

File: Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Components/PrimitiveMeshComponent.cs [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+namespace ECS.Unity.PrimitiveRenderer.Components
+{
+    public struct PrimitiveMeshComponent : IPoolableComponentProvider
+    {
+        public Mesh Mesh;
+        public PBMeshRenderer.MeshOneofCase SDKType;
+        public object PoolableComponent => Mesh;
+        public Type PoolableComponentType => typeof(Mesh);
+
+        public void Dispose() { }
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Components/PrimitiveMeshComponent.cs:15]: Let's move the empty `Dispose` implementation to the interface itself

**Review by mikhail-dcl — CHANGES_REQUESTED** [review]

File: Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs (hunk 8 above)
Comment (dalkia) [reply Explorer/Assets/Scripts/ECS/Unity/PrimitiveRenderer/Systems/InstantiatePrimitiveRenderingSystem.cs:86]: Such a silly mistake. Thanks for this finding

**Review by mikhail-dcl — APPROVED** [review]

**Pull request merged**